### ISSUE TITLE
Extra HTTP requests

### ISSUE THREAD

**Issue by Vlad-Shevliakov** [i18next/i18next-http-backend#48]
## 🐛 Bug Report

After loading the web application, an http request goes out to get localization data from *```loadPath```.
If you change the current language, then the corresponding files will come from the backend - this behavior is quite logical,because I only get files that will use.

Folders structure
```
- public/
--- locales/
----- ru
------- translation.json
------- filters.json
----- ua
------- translation.json
------- filters.json
```

![correct](https://user-images.githubusercontent.com/40333036/108237538-b00e9480-7150-11eb-81b1-7b7b46ee4330.png)

In i18n options I also use ```LanguageDetector```, which helps me to save the current state of the language in local storage (init file below)

```js
i18n
  .use(Backend)
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    initImmediate: false,
    fallbackLng: 'ru',
    debug: isDebugMode,
    ns: ['filters'],
    defaultNS: 'translation',
    interpolation: {
      escapeValue: false,
      formatSeparator: ','
    },
    react: {
      wait: true,
      useSuspense: true
    },
    backend: {
      loadPath: './locales/{{lng}}/{{ns}}.json' // * where i store my locales
    },
    detection: {
      order: ['localStorage'],
      caches: ['localStorage', 'cookie'],
      cookieMinutes: 80
    }
  })
```
Further, after switching the language and reloading the page, I will receive a `json` that corresponds to the current language. **But** there will also be extra requests to get localization files that should not be used

![wrong](https://user-images.githubusercontent.com/40333036/108237662-cd436300-7150-11eb-9d97-a67b2fc42a1f.png)
*red dots mark unnecessary requests

## To Reproduce

In other words.
I ran the page for the first time and got the files corresponding to 'fallbackLng' - from `locales/ru/*`
Then I changed the language to _ua_, the changes were saved in localStorage, and after reboot, **I hope** to only get the files from `locales/ua/*`.
But in the end I get files from both directories

## Expected behavior

After changing the language and **reloading** the page, I want to receive those files that **correspond** to the current language, avoiding unnecessary requests

## Your Environment

- *runtime version*: browser -> react-i18next v11.3.4
- *i18next version*: 19.3.4
- *os*: Windows / Mac

**Comment by adrai** [comment]
Check the url of the double requests, I'm pretty sure they're not the same.
Probably one is ru and the other ru-RU

**Comment by Vlad-Shevliakov** [comment]
@adrai In the video everything is as I described

https://user-images.githubusercontent.com/40333036/108243075-5610cd80-7156-11eb-823c-920ae55781e9.mp4

**Comment by adrai** [comment]
If you set debug to true, you'll probably see, that ru is loaded anyway, because it's your fallbackLng and because the languageDetector detected your  language ua, it will change the language to ua and load it accordingly.

**Comment by Vlad-Shevliakov** [comment]
@adrai 
`fallbackLng `is set to avoid errors in i18n. But this is not important either, because I will receive data from the path
 `http://localhost:3000/locales/dev/*` if you do not specify it
![image](https://user-images.githubusercontent.com/40333036/108244664-1b0f9980-7158-11eb-97bd-bf90e9658485.png)

**Comment by adrai** [comment]
the fallbackLng will always be loaded

**Comment by adrai** [comment]
The fallbackLng content is used, if there is a missing key in the current language

**Comment by Vlad-Shevliakov** [comment]
Yes. And it must be specified to avoid such an error:
![image](https://user-images.githubusercontent.com/40333036/108245254-ba349100-7158-11eb-9d80-3ef0ebe1acd8.png)
It is impossible not to specify it at the first start

But there is still an additional request to `http://localhost:3000/locales/dev/*` if you do not specify it.
I have no questions for this, the question is only in unnecessary http requests, it turns out that the backend does not take into account the detector data if there is one

**Comment by adrai** [comment]
ua comes from your detector and ru because it's the fallback

if you have other issues I would suggest you create a reproducable codesandbox example

**Comment by Vlad-Shevliakov** [comment]
The documentation has a great example
https://github.com/i18next/react-i18next/tree/master/example/react

If you run this, it will take the same behavior in the described problem :)

**Comment by adrai** [comment]
And this is the expected behaviour.
fallbackLng is always loaded

**Comment by adrai** [comment]
btw: if you don't like to have a fallbackLng... try to set it explictly to false

**Comment by Vlad-Shevliakov** [comment]
Let me try to describe the issue from a different angle.
That's right, it always boots, but in the case of a cold start
anyway, unnecessary requests occur.

It is strange why i18next-http-backend does not consider data received from the detector that was previously saved.

So it conflicts with the whole i18next-http-backend, try to run the example with clean cache and cleaned data in local storage. And you will get an **error** at every `t("transl.example")`

**Comment by adrai** [comment]
> That's right, it always boots, but in the case of a cold start

What is a cold start?

>  i18next-http-backend does not consider data received from the detector that was previously saved

There is no data received from the detector, the detector only saves the language (not the content)
i18next-http-backend is just a backend plugin, it does not know when to load what and why... i18next is responsible to trigger the backend plugin. Btw. if your backend sets some Control-Cache headers the browser will cache them automatically.

> try to run the example with clean cache and cleaned data in local storage. And you will get an error at every t("transl.example") without fallbackLng

I don't get any error...
![image](https://user-images.githubusercontent.com/1086194/108252918-78581a80-7159-11eb-9c35-636a53f8f386.png)
![image](https://user-images.githubusercontent.com/1086194/108252953-86a63680-7159-11eb-9a25-943b167ed8bb.png)

**Comment by adrai** [comment]
https://www.i18next.com/principles/fallback
![image](https://user-images.githubusercontent.com/1086194/108253243-d84ec100-7159-11eb-99f8-860456714e71.png)

**Comment by Vlad-Shevliakov** [comment]
I have read all related documentations
cold start - by these words I mean that you start the site without a cache and there is no data in local storages (cookie localStorage). For example, the very first launch of the application.

And if `fallbackLng` is not specified or set to false. You will get t('**title**') in html tag (for example) and not its value.
`fallbackLng: false`- solves the problem in half
Maybe my English is to blame? I don't know how to explain it otherwise :(

**Comment by adrai** [comment]
Sorry, I can't help you more than that... if setting `fallbackLng = false` does not help, I don't see anything else I can do...
All the examples I checked so far, work as expected and loads accordingly.
So by default i18next always loads the fallbackLng no matter if the current language is different or not.
The general behaviour I tried to explain here:

> There is no data received from the detector, the detector only saves the language (not the content)
i18next-http-backend is just a backend plugin, it does not know when to load what and why... i18next is responsible to trigger the backend plugin. Btw. if your backend sets some Control-Cache headers the browser will cache them automatically.

I'm sorry.

**Comment by Vlad-Shevliakov** [comment]
Ok, last chance.
If you have the opportunity, try the following steps:

- Set `fallbackLng: false`
- Run app
- Open DevTools then go to `'Application`' -> `'Storage'` and pres "Clear Site data" button.
*This is how you simulate the very first connection
- Reload the page
- See if all `i18n.t()` are working correctly?

Btw. After switching the language, you really get only the files you need that correspond to the current `{lng}` from `loadPath` without unnecessary http requests.

**Comment by adrai** [comment]
all fine, everything working correctly

**Comment by Vlad-Shevliakov** [comment]
Well. I will reinstall all my libraries, re-read the documentation again and try to figure out the cause of this problem. Perhaps this problem is from my side.
Probably should close this problem?

Thank you very much for your help and for 3 hours of your time!

**Comment by Vlad-Shevliakov** [comment]
I solved my problem ✨✨✨.
If `fallbackLng` is not specified or is false, then the `i18n.language` variable will be _undefined_, which in turn will result in invalid data in `i18n.t()`.
And this is the correct behavior inherent in the _react-i18next_ library.

Solve the problem of `i18n.t()` and unnecessary requests by triggering the `i18n.changeLanguage()` function from outside. Thus, we overwrite the value undefined and based on this, the back-end will receive the code of the **current** language and give only the necessary localization files.

In this case, the i18n options must contain:

```js
i18n
  .use(Backend)
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    initImmediate: false,
    fallbackLng: false,
 })
```
Then, at the top level of the application, you can set the value of the `i18n.language` variable.
```js
useEffect(() => {
    if (!i18n.language) {
      i18n.changeLanguage(lang)
      return
    }

    setCurrent(i18n.language)
}, [i18n.language])
```

**Comment by adrai** [comment]
So this issue can be closed now?

**Comment by Vlad-Shevliakov** [comment]
I think yes
